pkg/streamingpromql/types/data.go

```diff
@@ -27,6 +28,50 @@ type InstantVectorSeriesData struct {
 	Histograms []promql.HPoint
 }
 
+type InstantVectorSeriesDataIterator struct {
+	data   InstantVectorSeriesData
+	fIndex int
+	hIndex int
+}
+
+func (i *InstantVectorSeriesDataIterator) Reset(data InstantVectorSeriesData) {
+	i.fIndex = 0
+	i.hIndex = 0
+	i.data = data
+}
+
+// Next returns either a float or histogram iterating through both sets of points.
+// It returns the next point with the lowest timestamp.
+// If h is not nil, the value is a histogram, otherwise it is a float.
+// If no more values exist ok is false.
+func (i *InstantVectorSeriesDataIterator) Next() (t int64, f float64, h *histogram.FloatHistogram, ok bool) {
+	if i.fIndex >= len(i.data.Floats) && i.hIndex >= len(i.data.Histograms) {
+		return 0, 0, nil, false
+	}
+
+	if i.fIndex >= len(i.data.Floats) {
+		point := i.data.Histograms[i.hIndex]
+		i.hIndex++
+		return point.T, 0, point.H, true
+	}
+
+	if i.hIndex >= len(i.data.Histograms) {
+		point := i.data.Floats[i.fIndex]
+		i.fIndex++
+		return point.T, point.F, nil, true
+	}
+
+	// NOTE: Floats and Histograms should never exist at the same timestamp, T.
+	if i.data.Floats[i.fIndex].T < i.data.Histograms[i.hIndex].T {
+		point := i.data.Floats[i.fIndex]
+		i.fIndex++
+		return point.T, point.F, nil, true
+	}
+	point := i.data.Histograms[i.hIndex]
+	i.hIndex++
+	return point.T, 0, point.H, true
```

Comment:
Would it be possible to merge these cases into the `if`s above? Then we would have "return the next float" in one place and "return the next histogram" in another, and there's no risk of them becoming out of sync

Reply:
it's a little convoluted. It'd have to be something like:

```
	if i.fIndex < len(i.data.Floats) && (i.hIndex >= len(i.data.Histograms) || i.data.Floats[i.fIndex].T < i.data.Histograms[i.hIndex].T) {
		// Return the next float
		point := i.data.Floats[i.fIndex]
		i.fIndex++
		return point.T, point.F, nil, true
	}

	// Return the next histogram
	point := i.data.Histograms[i.hIndex]
	i.hIndex++
	return point.T, 0, point.H, true

```

Is that actually preferable in terms of readability?

Reply:
What if we extracted the `i.fIndex >= len(i.data.Floats)` and `i.hIndex >= len(i.data.Histograms)` expressions to variables like `exhaustedFloats` and `exhaustedHistograms`? Then all of the conditions in this method should become clearer - eg. this one becomes `!exhasutedFloats && (exhaustedHistograms || i.data.Floats[i.fIndex].T < i.data.Histograms[i.hIndex].T)`.